Assertion error `argextype only works on argument-position values` in optimizer when testing Checkpointing package on master

https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/f4801ff_vs_7b39515/Checkpointing.primary.log

```
Internal error: encountered unexpected error in runtime:
AssertionError(msg="argextype only works on argument-position values")
argextype at ./compiler/optimize.jl:371
argextype at ./compiler/optimize.jl:353 [inlined]
argextype at ./compiler/optimize.jl:353 [inlined]
stmt_effect_flags at ./compiler/optimize.jl:220
finish at ./compiler/optimize.jl:433
optimize at ./compiler/optimize.jl:523 [inlined]
_typeinf at ./compiler/typeinfer.jl:257
typeinf at ./compiler/typeinfer.jl:213
typeinf_edge at ./compiler/typeinfer.jl:914
abstract_call_method at ./compiler/abstractinterpretation.jl:605
abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:166
abstract_call_known at ./compiler/abstractinterpretation.jl:1744
abstract_call at ./compiler/abstractinterpretation.jl:1813
abstract_call at ./compiler/abstractinterpretation.jl:1792
abstract_eval_statement at ./compiler/abstractinterpretation.jl:1934
abstract_eval_basic_statement at ./compiler/abstractinterpretation.jl:2284
typeinf_local at ./compiler/abstractinterpretation.jl:2461
typeinf_nocycle at ./compiler/abstractinterpretation.jl:2559
_typeinf at ./compiler/typeinfer.jl:230
typeinf at ./compiler/typeinfer.jl:213
typeinf_ext at ./compiler/typeinfer.jl:1038
typeinf_ext_toplevel at ./compiler/typeinfer.jl:1071
typeinf_ext_toplevel at ./compiler/typeinfer.jl:1067
jfptr_typeinf_ext_toplevel_11977 at /Users/noemivitos/juliamaster/usr/lib/julia/sys.dylib (unknown line)
_jl_invoke at /Users/noemivitos/juliamaster/src/gf.c:2374
ijl_apply_generic at /Users/noemivitos/juliamaster/src/gf.c:2575
jl_apply at /Users/noemivitos/juliamaster/src/./julia.h:1846 [inlined]
jl_type_infer at /Users/noemivitos/juliamaster/src/gf.c:317
jl_generate_fptr_impl at /Users/noemivitos/juliamaster/src/jitlayers.cpp:345
```

Reduced to:

```julia
function foo()
    cglobal((a, ))
    ccall(0, Cvoid, (Nothing,), b)
end
```

```
julia> code_typed(foo, ())
ERROR: LoadError: AssertionError: argextype only works on argument-position values
Stacktrace:
  [1] argextype(x::Any, src::Core.Compiler.IRCode, sptypes::Vector{Any}, slottypes::Vector{Any})
    @ Core.Compiler ./compiler/optimize.jl:341
  [2] argextype (repeats 2 times)
    @ ./compiler/optimize.jl:323 [inlined]
  [3] stmt_effect_flags(lattice::Core.Compiler.OptimizerLattice, stmt::Any, rt::Any, src::Core.Compiler.IRCode)
    @ Core.Compiler ./compiler/optimize.jl:249
  [4] finish(interp::Core.Compiler.NativeInterpreter, opt::Core.Compiler.OptimizationState, params::Core.Compiler.OptimizationParams, ir::Core.Compiler.IRCode, caller::Core.Compiler.InferenceResult)
    @ Core.Compiler ./compiler/optimize.jl:403
  [5] optimize
    @ ./compiler/optimize.jl:493 [inlined]
  [6] _typeinf(interp::Core.Compiler.NativeInterpreter, frame::Core.Compiler.InferenceState)
    @ Core.Compiler ./compiler/typeinfer.jl:259
  [7] typeinf(interp::Core.Compiler.NativeInterpreter, frame::Core.Compiler.InferenceState)
    @ Core.Compiler ./compiler/typeinfer.jl:215
  [8] typeinf_frame(interp::Core.Compiler.NativeInterpreter, method::Method, atype::Any, sparams::Core.SimpleVector, run_optimizer::Bool)
    @ Core.Compiler ./compiler/typeinfer.jl:1003
  [9] typeinf_code(interp::Core.Compiler.NativeInterpreter, method::Method, atype::Any, sparams::Core.SimpleVector, run_optimizer::Bool)
    @ Core.Compiler ./compiler/typeinfer.jl:955
 [10] code_typed_by_type(tt::Type; optimize::Bool, debuginfo::Symbol, world::UInt64, interp::Core.Compiler.NativeInterpreter)
    @ Base ./reflection.jl:1271
 [11] #code_typed#24
    @ ./reflection.jl:1230 [inlined]
 [12] code_typed
    @ ./reflection.jl:1221 [inlined]
```

--- a/base/compiler/optimize.jl
+++ b/base/compiler/optimize.jl
@@ -243,6 +243,10 @@ function stmt_effect_flags(lattice::AbstractLattice, @nospecialize(stmt), @nospe
                 nothrow = _builtin_nothrow(lattice, f, argtypes, rt)
                 return (true, nothrow, nothrow)
             end
+            if f === Intrinsics.cglobal
+                # TODO: these are not yet linearized
+                return (false, false, false)
+            end
             isa(f, Builtin) || return (false, false, false)
             # Needs to be handled in inlining to look at the callee effects
             f === Core._apply_iterate && return (false, false, false)

--- a/base/compiler/ssair/verify.jl
+++ b/base/compiler/ssair/verify.jl
@@ -39,7 +39,6 @@ function check_op(ir::IRCode, domtree::DomTree, @nospecialize(op), use_bb::Int, 
         else
             if !dominates(domtree, def_bb, use_bb) && !(bb_unreachable(domtree, def_bb) && bb_unreachable(domtree, use_bb))
                 # At the moment, we allow GC preserve tokens outside the standard domination notion
-                #@Base.show ir
                 @verify_error "Basic Block $def_bb does not dominate block $use_bb (tried to use value $(op.id))"
                 error("")
             end
@@ -62,7 +61,6 @@ function check_op(ir::IRCode, domtree::DomTree, @nospecialize(op), use_bb::Int, 
             end
         end
     elseif isa(op, Union{OldSSAValue, NewSSAValue})
-        #@Base.show ir
         @verify_error "Left over SSA marker"
         error("")
     elseif isa(op, Union{SlotNumber, TypedSlot})
@@ -245,6 +243,12 @@ function verify_ir(ir::IRCode, print::Bool=true, allow_frontend_forms::Bool=fals
                         (stmt.args[1] isa GlobalRef || (stmt.args[1] isa Expr && stmt.args[1].head === :static_parameter))
                     # a GlobalRef or static_parameter isdefined check does not evaluate its argument
                     continue
+                elseif stmt.head === :call
+                    f = stmt.args[1]
+                    if f isa GlobalRef && f.name === :cglobal
+                        # TODO: these are not yet linearized
+                        continue
+                    end
                 end
             end
             n = 1

--- a/test/compiler/inference.jl
+++ b/test/compiler/inference.jl
@@ -4233,3 +4233,14 @@ function unused_apply_iterate()
     return nothing
 end
 @test fully_eliminated(unused_apply_iterate, ())
+
+@testset "#45956: non-linearized cglobal needs special treatment for stmt effects" begin
+    function foo()
+        cglobal((a, ))
+        ccall(0, Cvoid, (Nothing,), b)
+    end
+    @test only(code_typed() do
+        cglobal((a, ))
+        ccall(0, Cvoid, (Nothing,), b)
+    end)[2] === Nothing
+end

--- a/test/compiler/ssair.jl
+++ b/test/compiler/ssair.jl
@@ -173,7 +173,17 @@ let ci = make_ci([
     ])
     ir = Core.Compiler.inflate_ir(ci)
     ir = Core.Compiler.compact!(ir, true)
-    @test Core.Compiler.verify_ir(ir) == nothing
+    @test Core.Compiler.verify_ir(ir) === nothing
+end
+
+# Test that the verifier doesn't choke on cglobals (which aren't linearized)
+let ci = make_ci([
+        Expr(:call, GlobalRef(Main, :cglobal),
+                    Expr(:call, Core.tuple, :(:c)), Nothing),
+                    Core.Compiler.ReturnNode()
+    ])
+    ir = Core.Compiler.inflate_ir(ci)
+    @test Core.Compiler.verify_ir(ir) === nothing
 end
 
 # Test that GlobalRef in value position is non-canonical

